Sources/StreamChat/StateLayer/ChatState+Internal.swift

```diff
@@ -0,0 +1,101 @@
+//
+// Copyright © 2024 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+@available(iOS 13.0, *)
+extension ChatState {
+    struct UnreadMessageLookup {
+        let userId: UserId
+        let readStates: [ChatChannelRead]
+        let messageOrder: MessageOrdering
+        let messages: StreamCollection<ChatMessage>
+        let hasLoadedAllOldestMessages: Bool
+        
+        @MainActor static func firstUnreadMessage(
+            in state: ChatState,
+            userId: UserId
+        ) -> MessageId? {
+            guard let channel = state.channel else { return nil }
+            let lookup = UnreadMessageLookup(
+                userId: userId,
+                readStates: channel.reads,
+                messageOrder: state.messageOrder,
+                messages: state.messages,
+                hasLoadedAllOldestMessages: state.hasLoadedAllOldestMessages
+            )
+            return lookup.firstUnreadMessageId
+        }
+        
+        private var firstUnreadMessageId: MessageId? {
+            guard let readInfo = readStates.first(where: { $0.user.id == userId }) else {
+                // Read state is unavailable
+                return hasLoadedAllOldestMessages ? oldestRegularMessageId : nil
+            }
+            guard readInfo.unreadMessagesCount > 0 else { return nil }
+            guard let lastReadMessageId = readInfo.lastReadMessageId else {
+                // Everything is unread if read state is there but there is no lastReadMessageId
+                return hasLoadedAllOldestMessages ? oldestRegularMessageId : nil
+            }
+            if let lastReadIndex = indexOfMessageId(lastReadMessageId) {
+                if isMostRecent(at: lastReadIndex) {
+                    // Everything has been read
+                    return nil
+                } else {
+                    return lookupUnreadMessageId(after: lastReadIndex)
+                }
+            } else {
+                // Can't reach the last read message (if all have been loaded then the channel might have been truncated or hidden, in that case, use the oldest message)
+                return hasLoadedAllOldestMessages ? oldestRegularMessageId : nil
+            }
+        }
```

Comment:
This is copy pasted from `ChatChannelController` right? 🤔 The problem here is that we now need 2 maintain the same logic twice which is a bit concerning

Reply:
I agree, we can probably introduce some helper class. We need to update the logic anyway soon, so having it in 2 places is a bit error prone.